File: 01_GOF23/DP_examples/Go/01_Creational/Factory/ConcreteCreator.go
package Factory

// ShapeFactory 形状工厂
type shapeFactory struct {
	factories map[ShapeKind]ICreator
}

var ShapeFactory shapeFactory = shapeFactory{
	factories: make(map[ShapeKind]ICreator),
}

func (f shapeFactory) GetShape(kind ShapeKind) IShape {
	creator := f.getFactory(kind)
	return creator.GetShape()
}

// GetFactory 根据类型获取创建者
func (f shapeFactory) getFactory(kind ShapeKind) ICreator {
	if _, ok := f.factories[kind]; !ok {
		f.factories[kind] = newCreator(kind)
	}
	return f.factories[kind]
}

File: 01_GOF23/DP_examples/Python/01_Creational/Factory/ConcreteCreator.py
from Creator import UnknownShapeCreator, CircleCreator, SquareCreator, RectangleCreator, BigCircleCreator
from Product import ShapeKind

# 未知形状创建者
# 形状工厂
class ShapeFactory:
    _factories = {}
    def get_shape(kind):
        creator = ShapeFactory._get_factory(kind)
        return creator.get_shape()
    
    def _get_factory(kind):
        if kind in ShapeFactory._factories:
            return ShapeFactory._factories[kind]
        
        if kind == ShapeKind.Circle:
            creator = CircleCreator()
        elif kind == ShapeKind.Square:
            creator = SquareCreator()
        elif kind == ShapeKind.Rectangle:
            creator = RectangleCreator()
        elif kind == ShapeKind.BigCircle:
            creator = BigCircleCreator()
        else:
            creator = UnknownShapeCreator()
        
        ShapeFactory._factories[kind] = creator
        return creator
